### PR TITLE
Preset date and time for sending invitation emails

### PR DESCRIPTION
**Pull request by pulkitkalra**
This pull request completes #26.

The feature added here is the ability to preset a date and time to send invitation emails.
User specifies if they want to send emails immediately or at preset time. This preset time is used to determine the delay in sending emails and then the invitation job is registered within the scheduler.

image

The emails are scheduled using Quartz.

## **Manual Testing in development**
- Create a new study (with a participant and an email you can access)
- From the study page, select **Preset Date and Time**
- Specify a time in the future: e.g. 5 mins after current time (Note: you must keep the debugging process running since the quartz server runs in that process).
- Check your email at the time you specified to verify that the email was received.

## **Input Validation**
- If a user selects a date/ time in the past, server side validation will ensure that the email is sent out immediately.
- Client side validation has not been implemented in this PR.

## **Dependencies**
- A quartz dependency was added to the pom file, there is a small chance that this may need to modified once #49 is merged in.
- Expecting unit test cases to be added by #36, likely to be some semantic conflicts from that.

### CODE REVIEW

**Comment by pulkitkalra** [comment]
@st970703 @victorlian please review thanks :)

**Comment by softeng-701** [comment]
@pulkitkalra Resolve conflicts, please!

File: src/main/java/org/project36/qualopt/service/StudyService.java [hunk 1]
```diff
@@ -75,4 +102,31 @@ protected PasswordAuthentication getPasswordAuthentication() {
...
+        messageMap.put("message", message);
+
+        JobDetail job = newJob(EmailScheduler.class)
+            .withIdentity("myJob", "group1")
+            .usingJobData(messageMap)
+            .build();
+
+        // Trigger the job to run after a certain delay
+        Trigger trigger = newTrigger().withIdentity("myTrigger", "group1")
+        .startAt(futureDate(delay,IntervalUnit.SECOND))
```
Comment (victorlian) [inline src/main/java/org/project36/qualopt/service/StudyService.java]: nitpicking on the indentation here.... :( sorry.

File: src/main/java/org/project36/qualopt/web/rest/StudyResource.java [hunk 2]
```diff
@@ -165,12 +167,14 @@ else if (authentication.getPrincipal() instanceof String) {
     @PostMapping(path = "/studies/send",
         produces={MediaType.APPLICATION_JSON_VALUE, MediaType.TEXT_PLAIN_VALUE})
     @Timed
-    public ResponseEntity sendStudy(@Valid @RequestBody Study study) {
-        log.debug("REST request to send Study : {}", study);
+    public ResponseEntity sendStudy(@Valid @RequestBody Invitation inv) {
+        Study study = inv.getStudy();
+        int delay = Integer.parseInt(inv.getDelay())/1000;
```
Comment (victorlian) [inline src/main/java/org/project36/qualopt/web/rest/StudyResource.java]: is there a specific reason why delay is of type string but not int? (maybe provide a comment if this is a work around for something?)

**Review by victorlian — CHANGES_REQUESTED** [review]
LGTM apart from several tiny issues

**Comment by victorlian** [comment]
Also lots of merge conflicts for you :(

File: src/main/java/org/project36/qualopt/web/rest/StudyResource.java (hunk 2 above)
Comment (st970703) [reply src/main/java/org/project36/qualopt/web/rest/StudyResource.java]: Not too sure about why inv.getDelay() returns a string that needs to be parsed?

File: src/main/webapp/app/entities/study/study-detail.component.ts [hunk 3]
```diff
@@ -25,11 +28,26 @@ export class StudyDetailComponent implements OnInit, OnDestroy {
...
         });
         this.registerChangeInStudies();
+        this.currentTimeStamp = StudyService.getCurrentDateTime();
+        this.currentTimeZone = this.getTimeZone();
+    }
+
+    // credit for function: https://stackoverflow.com/questions/1091372/getting-the-clients-timezone-in-javascript
+    // User: Mr_Green
+    getTimeZone(): string {
+        var offset = new Date().getTimezoneOffset(), o = Math.abs(offset);
```
Comment (st970703) [inline src/main/webapp/app/entities/study/study-detail.component.ts]: Try using const and let in TS. It is not a good idea to mix JS into TS.
'var offset'

File: src/main/webapp/app/entities/study/study-detail.component.ts [hunk 4]
```diff
@@ -25,11 +28,26 @@ export class StudyDetailComponent implements OnInit, OnDestroy {
...
+        return (offset < 0 ? "+" : "-") + ("00" + Math.floor(o / 60)).slice(-2) + ":" + ("00" + (o % 60)).slice(-2);
```
Comment (st970703) [inline src/main/webapp/app/entities/study/study-detail.component.ts]: That's quite a long return line?
I will store that in a return variable personally

File: src/main/java/org/project36/qualopt/domain/Invitation.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import org.project36.qualopt.domain.Study;
+/**
+ * The invitation class is a domain class that is NOT MAPPED
+ * and is used to pass a study object along with the delay to
+ * the resource method.
+ */
+public class Invitation implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private Study study;
```
Comment (st970703) [inline src/main/java/org/project36/qualopt/domain/Invitation.java]: According to Ewan's Java rules, you should add underscores in front of the private fields?

File: src/main/java/org/project36/qualopt/service/StudyService.java (hunk 1 above)
Comment (pulkitkalra) [reply src/main/java/org/project36/qualopt/service/StudyService.java]: this has been fixed now :)

File: src/main/java/org/project36/qualopt/web/rest/StudyResource.java (hunk 2 above)
Comment (pulkitkalra) [reply src/main/java/org/project36/qualopt/web/rest/StudyResource.java]: thank you for pointing that out, changed to int

File: src/main/webapp/app/entities/study/study-detail.component.ts (hunk 3 above)
Comment (pulkitkalra) [reply src/main/webapp/app/entities/study/study-detail.component.ts]: thanks, fixed.

File: src/main/webapp/app/entities/study/study-detail.component.ts (hunk 4 above)
Comment (pulkitkalra) [reply src/main/webapp/app/entities/study/study-detail.component.ts]: hmm, then the variable assignment would be a long line?
I got that function from SO, which why I didn't modify it

File: src/main/java/org/project36/qualopt/domain/Invitation.java (hunk 5 above)
Comment (victorlian) [reply src/main/java/org/project36/qualopt/domain/Invitation.java]: I disagree to this one :( 
The existing code base doesn't use the underscore convention, so any code added on top of it probably shouldn't use it either.

Comment (pulkitkalra) [reply src/main/java/org/project36/qualopt/domain/Invitation.java]: that's more of a convention isn't it? Plus this project isn't following that, so I didn't put the underscore in

**Comment by wilmol** [comment]
Can you run StudyResource sendStudy test.

**Comment by pulkitkalra** [comment]
@wilmol ahhh thank you for pointing that out, looks like I missed that test. The test was failing because the post request was sending a study object but now needs to send an invitation object instead. This has been fixed now :)

**Review by victorlian — APPROVED** [review]
I have tested this and can verify that the email will be sent on time :) 
Code LGTM 👍

**Review by st970703 — APPROVED** [review]

**Comment by pulkitkalra** [comment]
@softeng-701 ready to be merged :)

**Comment by softeng-701** [comment]
@pulkitkalra Squash please!

**Comment by softeng-701** [comment]
@pulkitkalra conflicts arose again, please check!

**Comment by pulkitkalra** [comment]
@softeng-701 I've resolved the conflicts, but squashing is not working due to the merge commits I have. I'm having to create a new branch and then will need to make a **new pull request** if there is a requirement to have a single commit.

**Comment by pulkitkalra** [comment]
@softeng-701 resolved and squashed. Ready to merge :)

**Comment by pulkitkalra** [comment]
@softeng-701 bump. This is ready to go

**Pull request merged**